### ISSUE TITLE
Enhancement: Read-only vertices property for shapes & sprites

### ISSUE THREAD

**Issue by pushfoo** [pyglet/pyglet#939]
## TL;DR

1. Add a read-only `vertices` property returning a value matching `Sequence[Point2D]` to any object which both:
   1.  has geometry / a bounding box
   2.  a user could reasonably want to perform collision against
2. Add a `HasVerticesProperty` [`Protocol` type]() which matches a `vertices` property, but not an `attribute`:
   ```python
   # Protocols are first available on 3.8, so this requires no external dependencies
   from typing import Protocol, TypeVar, Sequence

   # Optional: allow specifying 2D or 3D vertices
   TPoint = TypeVar('TPoint')

   class HasVerticesProperty(Protocol[TPoint]):
        """Generic annotation for anything object with a vertices property"""

        @property
        def vertices(self) -> Sequence[TPoint]:
            ...

    ```

3. Get huge benefits:
    1. Colliders can be simpler since they can delegate data storage & vertex recalculation
    2. Sprites and shapes are now their own default hitbox data
    3. You can still generate and use better hitbox data as needed

## Why?

1. First, it simplifies implementing collisions (including #928) by allowing colliders to delegate recalculation of vertices to other objects:
    ```python
    class PolygonCollider2D:
       
        def __init__(self, data_source: Optional[HasVerticesProperty[Point2D]):
           """Create a polygon collider which delegates data reads and method calls
          
           Args:
               data_source: An object to delegate method calls to
           """
           self._data_source = data_source

        @property
        def vertices(self) -> Sequence[Point2D]:
           return self._data_source.vertices

        def collides_with(self, other: HasVerticesProperty[Point2D]) -> bool:
              if have_optimized_collision(self, other):  # psuedocode, I'm stll working on optimizing this
                  return self.optimized_collision(self, other)
     
              return separating_axis_theorem(self.vertices, other.vertices)

        # various properties omitted, but we'll probably still need stuff like position and anchor
    ```
    1. Since the base polygon collider stores a reference to a vertex source object responsible for recalculation:
        1. We do not have to duplicate code between the collider and the shapes
        2. We do not have to deal with syncing or recalculating data from shapes to colliders
    2. Unoptimized shape-specific collider subclasses can be reduced to `__init__` / class methods which take either:
        1. An instance of a drawable shape, in which case a reference would be stored as above
        3. The same arguments as the corresponding shape type, in which case a new instance would be stored and used as the source

2. Sprites and shapes are now also usable as collidables / hitboxes
3. Adding debug drawing features to collider objects is now much easier

### Implementation steps

- [ ] Add a `HasVerticesProperty` Protocol type which matches a property getter returning `Sequence[TPoint]` or `Sequence[Point2D]`
- [ ] Either of the following, or a better equivalent:

     * Store tuples of vertices on shapes
     * Add a `ReadOnlyVertexView` class which:
       1. Matches`Sequence[Point2D]`
       2. Returns `Point2D` on demand for indices

- [ ] Add a matching property to the following types:
     - [ ] `Sprite`
     - [ ] `ShapeABC`
     - [ ] Custom versions of the property for anything which requires it

### Assumptions

1. People are passing closed polygons / we don't need to validate this
2. Delegation's performance impact is worth the increased flexibility
3. We don't need to support instance or class attributes as vertex sources

### Why not attributes?

1. Type checkers treat attributes and properties as different things
    1. Supporting both requires verbose boilerplate + union types
    2. Potential for combinatorial explosion among annotations and their componenets

2. It would match the `Glyph` class here:
        https://github.com/pyglet/pyglet/blob/1af5a55f7aedac4a5245e7a306bb1c7cc1bdeda8/pyglet/font/base.py#L124

3. People do not seem interested in using attributes

## Future areas to explore (if needed)

* [ctypes struct array views](https://docs.python.org/3/library/ctypes.html#arrays) of the same underlying position attribute data. 
* Using `memoryview.toreadonly` since it works with floats (this could require modifying vertex lists)

**Comment by pushfoo** [comment]
Following up on a [Discord discussion with caffienepills](https://discord.com/channels/438622377094414346/477394331112701952/1152359934487445594):

1. The internally stored vertex positions for some objects aren't necessarily final coordinates
2. Shaders may perform additional translation, rotation, and scaling
3. This property would have to account for that

**Comment by caffeinepills** [comment]
Revisiting this, but honestly I don't think this will realistically be possible. 

As discussed before, since a shader can manipulate the vertices/position of things, we can't realistically get a 1:1 of the vertices being on the screen. A shader can move things over time, different translations, scaling. If someone sets their own custom shader, then they will quickly become invalid properties.

We could offer some very basic left, top, bottom, right properties like I did with `TextLayout`'s. However, this is of course limiting as well, in that it does not take rotation or other shader changes into account. If that's still sufficient enough, let us know, and I can add that.

The point of moving things off of the CPU was to take advantage of the GPU, so having CPU calculate the vertices again seems like taking a big step back. I can't really think of a way to make this work other than what was stated.

**Comment by pushfoo** [comment]
TL;DR: Even basic lrtb properties would be great.

> We could offer some very basic left, top, bottom, right properties like I did with `TextLayout`'s

Yes, please. This ticket tried to break #928 into more manageable components and deal with assumptions I saw as risky. These include:

* The assumption that the viewport / rendering will be axis aligned
* Mixing collision and rendering into shapes

> Going back to having CPU calculate the vertices again seems like taking a big step back.

In an ideal world, the point storage objects might have both CPU and GPU based versions which can pass data back to Python. I'm not sure how feasible this is since it's been a few months since I looked at our vertex domain code. I know it's possible to get data back to Python from VRAM, but I'm not sure if requires compute shaders.

> If someone sets their own custom shader, then they will quickly become invalid properties.

Delegation of points and their recalculation was a way of dealing with this. It gives the user a way to make up for breaking the assumptions of our built-ins. The bigger issue is that this might take a lot of work to implement.

**Comment by pushfoo** [comment]
TL;DR: Rotation-sensitive w/h/lrtb might be more generally useful than we thought

I brought up this thread with others recently in the context of games. Although rotation-sensitive width/height/lrtb seemed like something only useful for world/view rotation as a game mechanic, that overlooks its uses in non-rectilinear UI and non-game applications.

In addition to radial menus which resize icons, it seems like there are a lot of opportunities in info overlay layouts such as:

* GIS / map rotation
* Summarizing and overlaying video data tagging, perhaps like the annotations I understand @matan3178 works on

To be sure I understand correctly, the only concept of cameras we have [is in the examples](https://github.com/pyglet/pyglet/blob/a9186de30e21d501686c519e64bb594fc2831c2d/examples/window/camera.py), right?
